doc/design/error_handling_requirements.md

```diff
@@ -0,0 +1,228 @@
+# Error Handling
+
+## Faults, Errors and Failures
+
+We use the following error terminology.
+
+1. A **fault** is a design flaw, a programming error or can e.g. be caused by unforseen use
+   (oversight of a use case is basically also a design flaw).
+2. An **error** is the result of a fault and manifests as an incorrect or inconsistent (intermediate) result.
+   An error may not be externally observable by the user of the system.
+3. A **failure** occurs if the system observably acts in contradiction to its specification.
+
+This reasonably assumes that errors are always caused by faults. If these errors are not detected
+they can cause failures. By detecting and handling errors failures are avoided.
+
+The goal of this document is to define error handling requirements.
+
+## Error Detection vs. Error Handling
+
+To handle an error it obviously needs to be detected first. To do so, the code must contain conditionals that check
+for inconsistent states, unavailability of resources, return codes of system calls etc.
+This means that the error has to be anticipated by the system designer,
+i.e. we cannot handle unforseen errors since they are not detected by definition.
+
+Error handling is the execution of handling code once the error was detected.
+This ranges from simple logging to sophisticated recovery or shutdown code and depends on the error in question.
+We distinguish between recoverable and non-recoverable error handling.
+
+### Recoverable Error Handling
+
+A recoverable error handling strategy allows the system to continue operation after the handling code
+was executed successfully. Exceptions are an example of such a mechanism but also any handler code
+that does not result in termination. If the handling code itself fails (e.g. insufficient resources)
+no recovery is possible and the system terminates.
+
+### Non-recoverable Error Handling
+
+A non-recoverable error handling strategy will always call terminate after the handling code was executed.
+Handler code may be used to initiate a graceful shutdown. An example are (strict) asserts.
+
+## Requirements
+
+In the following we list the requirements of the Error Handling (EH) strategy. The term platform
+refers to the usage environment of the error handler. There can be various platforms,
+e.g. for regular operations, testing, different OS etc.
+
+### Functional
+
+1. The EH shall support error codes.
+    - configurable on a platform basis
+    - errors from different platforms shall not be usable (checked at compile time)
+    - uniqueness is not enforced (can be done by the layer above if desired)
+    - error codes shall be optional (if none is used, it is mapped to some default error code)
+1. The EH shall support multiple error levels.
+    - the individual error levels are defined by the platform
```

Comment:
The error levels must be defined by the error handler and non configurable. Otherwise we only can use the error level `FATAL` in our code base since we never know if some platform may not have this specific error level.

Otherwise I think we can get rid completely of error levels when we state the error handler is only called on non recoverable errors.